### PR TITLE
fix #35 Fixed Initial State Behavior

### LINKED ISSUE #35: Don't process initial state until it is needed
Rather than computing initial states during block configuration, we should be evaluating the expressions when the signal comes in. This will allow us to do the following things:

- Use attributes on incoming signals to determine initial state
- Use time based expressions in initial state that are tied to when the first signal came through

### PR DESCRIPTION
**Pull request by kevincrabbe**
Now the initial state is not set until a signal passes through the block. Wrote test to test when if initial state is set before or after a signal passes through.

### CODE REVIEW

File: tests/test_state_base.py [hunk 1]
```diff
@@ -118,3 +118,20 @@ def test_current_state_command(self):
             sorted(blk.current_state(group=None), key=lambda k: k['group']),
                    [{'group': 'A', 'state': 2}, {'group': 'B', 'state': 1}])
         blk.stop()
+
+    def test_process_initial_state(self):
+        """ Test that the initial state gets
```
Comment (hansmosh) [inline tests/test_state_base.py]: Take another look at PEP-257 about mult-line docstrings

**Comment by hansmosh** [comment]
This solution still misses one of the features requested by issue #35 

> Use attributes on incoming signals to determine initial state

The initial state should be run against the incoming signal that triggered the state change, not just a new empty `Signal()`. `StateBase.get_state` may need to be updated to have an optional argument `signal`.

File: state_base.py [hunk 2]
```diff
@@ -48,6 +47,8 @@ def get_state(self, group):
         initial state configured for the block. It will also set that as the
         state.
         """
+        # initial state is set once a signal passes through
+        self._initial_state = self.initial_state(Signal())
         if group not in self._states:
             self._states[group] = copy(self._initial_state)
```
Comment (hansmosh) [inline state_base.py]: I can't recall what this `copy` is here for but I'm thinking it's not necessary anymore based on the new code. We probably don't even need the `self._initial_state` attribute

**Comment by hansmosh** [comment]
@kevincrabbe: A couple comments for you

**Comment by hansmosh** [comment]
Take another look at good commit messages too: http://chris.beams.io/posts/git-commit/

It would look cleaner as something more simple, like:

> Initial state set when first signal is processed
>
> fix #35

File: state_base.py [hunk 3]
```diff
@@ -48,8 +46,9 @@ def get_state(self, group):
         initial state configured for the block. It will also set that as the
         state.
         """
+        # initial state is set once a signal passes through
```
Comment (hansmosh) [inline state_base.py]: Remove this comment

File: state_base.py [hunk 4]
```diff
@@ -48,8 +46,9 @@ def get_state(self, group):
         initial state configured for the block. It will also set that as the
```
Comment (hansmosh) [inline state_base.py]: Update this to reflect what the block is actually doing now

File: tests/test_state_base.py [hunk 5]
```diff
@@ -118,3 +118,19 @@ def test_current_state_command(self):
...
+                'initial_state': '{{5}}',
+                'group_by': '{{group}}'
+            })
+        blk.start()
+        # make sure that the initial state has not been set
+        self.assertEqual(blk._states, {})
+
+        #now make sure that the intial state has been set after a signal has passed through
+        self.assertEqual(blk.get_state('A'), 5)
+        self.assertEqual(blk._states['A'], 5)
```
Comment (hansmosh) [inline tests/test_state_base.py]: This test doesn't pass

**Comment by hansmosh** [comment]
@kevincrabbe, back at you. The new unit test doesn't pass.

**Comment by hansmosh** [comment]
We discussed last week that maybe we shouldn't even be making this change. I proposed that we discuss this further with Matt. You started that on issue #35 but never followed up on what he had to see. Please do that.

If you guys come to the conclusion that we don't need this change, please close out this PR without merging it.

**Comment by kevincrabbe** [comment]
We are not going forward with this change.

**Pull request closed without merging**